tests/beta/tapes/test_qnode.py

```diff
@@ -0,0 +1,429 @@
+
+        # test an unknown model raises an exception
+
+        def capabilities(cls):
+            capabilities = cls._capabilities
+            capabilities.update(model="None")
+            return capabilities
+
+        monkeypatch.setattr(qml.devices.DefaultQubit, "capabilities", capabilities)
+        dev = qml.device("default.qubit", wires=1)
+
+        with pytest.raises(QuantumFunctionError, match="does not support the parameter-shift rule"):
+            QNode._get_parameter_shift_method(dev, None)
+
+    def test_best_method(self, monkeypatch):
+        """Test that the method for determining the best diff method
+        for a given device and interface works correctly"""
+        dev = qml.device("default.qubit", wires=1)
+        monkeypatch.setitem(dev._capabilities, "passthru_interface", 1)
+        monkeypatch.setitem(dev._capabilities, "provides_jacobian", True)
+
+        # backprop is given priority
+        res = QNode._get_best_method(dev, 1)
+        assert res == (QuantumTape, None, "backprop")
+
+        # device is the next priority
+        res = QNode._get_best_method(dev, 2)
+        assert res == (QuantumTape, 2, "device")
+
+        # The next fallback is parameter-shift.
+        monkeypatch.setitem(dev._capabilities, "provides_jacobian", False)
+
+        # TODO: remove this monkeypatching once the parameter shift tape is implemented
+        with monkeypatch.context() as m:
+            m.setattr(QNode, "_get_parameter_shift_method", lambda d, i: ("QubitParamShiftTape", i, "analytic"))
+            res = QNode._get_best_method(dev, 2)
+            assert res == ("QubitParamShiftTape", 2, "analytic")
+
+        # finally, if both fail, finite differences is the fallback
+        def capabilities(cls):
+            capabilities = cls._capabilities
+            capabilities.update(model="None")
+            return capabilities
+
+        monkeypatch.setattr(qml.devices.DefaultQubit, "capabilities", capabilities)
+        res = QNode._get_best_method(dev, 2)
+        assert res == (QuantumTape, 2, "numeric")
+
+    def test_diff_method(self, mocker):
+        """Test that a user-supplied diff-method correctly returns the right
+        quantum tape, interface, and diff method."""
+        dev = qml.device("default.qubit", wires=1)
+
+        mock_best = mocker.patch("pennylane.beta.tapes.QNode._get_best_method")
+        mock_best.return_value = 1, 2, 3
+
+        mock_backprop = mocker.patch("pennylane.beta.tapes.QNode._validate_backprop_method")
+        mock_backprop.return_value = 4, 5, 6
+
+        mock_device = mocker.patch("pennylane.beta.tapes.QNode._validate_device_method")
+        mock_device.return_value = 7, 8, 9
+
+        qn = QNode(None, dev, diff_method="best")
+        assert qn._tape == mock_best.return_value[0]
+        assert qn.interface == mock_best.return_value[1]
+        assert qn.diff_options["method"] == mock_best.return_value[2]
+        mock_best.assert_called_once()

```

Comment:
Could the `assert_called_once()` go at the end, to make sure the `_get_best_method()` wasn't called for the cases below?

Reply:
done!